dev/bus/export/dev/serial.hpp

```diff
@@ -0,0 +1,286 @@
+    static uint8_t m_tx_buffer[buf_size];
+
+    static std::atomic_bool m_rx_new_xfer_allowed;
+    static binary_semaphore m_rx_is_data_available;
+
+    static std::atomic_size_t m_rx_write_iter; //!< Used for accepting data from the bus
+    static std::atomic_size_t m_rx_read_iter; //!< Used for returning data to the user
+
+    static bool m_is_inited;
+};
+
+template <class PBus, size_t buf_size>
+uint8_t serial<PBus, buf_size>::m_rx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+uint8_t serial<PBus, buf_size>::m_tx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_write_iter;
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_read_iter;
+
+template <class PBus, size_t buf_size>
+bool serial<PBus, buf_size>::m_is_inited;
+
+template <class PBus, size_t buf_size>
+std::atomic_bool serial<PBus, buf_size>::m_rx_new_xfer_allowed;
+
+template <class PBus, size_t buf_size>
+binary_semaphore serial<PBus, buf_size>::m_rx_is_data_available;
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::init()
+{
+    ecl_assert(!m_is_inited);
+    auto result = PBus::init();
+    if (is_error(result)) {
+        return result;
+    }
+    PBus::set_handler(bus_handler);
+    result = PBus::enable_listen_mode();
+    if (is_error(result)) {
+        return result;
+    }
+    m_rx_new_xfer_allowed = true;
+    PBus::set_rx(m_rx_buffer, buffer_size);
+    result = PBus::do_xfer();
+    if (is_ok(result)) {
+        m_is_inited = true;
+    }
+    return result;
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::deinit()
+{
+    ecl_assert(m_is_inited);
+    PBus::cancel_xfer();
+    PBus::reset_buffers();
+    m_rx_read_iter = 0;
+    m_rx_write_iter = 0;
+    m_is_inited = false;
+    return err::ok;
+}
+
+template <class PBus, size_t buf_size>
+void serial<PBus, buf_size>::bus_handler(bus_channel ch, bus_event type, size_t total)
+{
+    if (ch == bus_channel::rx) {
+        ecl_assert(type == bus_event::tc);
+        m_rx_write_iter++;
+        ecl_assert(m_rx_write_iter == total);
+
+        // The buffer has been overflowed.
+        // Move read iterator to point to the most relevant data
+        //
+        // It might be a bit unclear why the same expression
+        // m_rx_read_iter == m_rx_write_iter both means that the buffer is
+        // overflown and empty (see recv_buf()).
+        // However, the context of bus_handler() implies that the
+        // read iterator is "approached" by write iterator from behind
+        // and we properly adjust the position of read iterator to be
+        // always ahead of the write iterator.
+        if (m_rx_read_iter == m_rx_write_iter) {
+            m_rx_read_iter++;
+            if (m_rx_read_iter == buffer_size) {
+                m_rx_read_iter = 0;
+            }
+        }
+
+        m_rx_is_data_available.signal();
```

Comment:
Let's call it only if there were no data before .
It possibly should be checked somewhere higher. Taking to the account our invariant that `m_rx_read_iter == m_rx_write_iter` only in case if there is no data left, we can signal semaphore using that check.

Avoiding excessive calls to `signal()` will prevent additional SVC calls and interrupts (in case RTOS is enabled)